PORTING_GUIDE.md

```diff
@@ -0,0 +1,174 @@
+Guidelines for porting WPILib examples to Python
+================================================
+
+To ensure that our examples are helpful and accurate for those learning how to
+use RobotPy, we have a set of guidelines for adding new examples to the project.
+These guidelines are not strictly enforced, but we do ask that you follow them
+when submitting pull requests with new examples. This will make the review
+process easier for everyone involved.
+
+In general, our examples are based on the Java examples from allwpilib, as Java
+is often easier to translate to Python. However, not all of our existing
+examples adhere to all of these guidelines. If you see an opportunity to improve
+an existing example, feel free to make the necessary changes.
+
+Shorter thoughts
+----------------
+
+Testing:
+
+* New examples must run! You *must* test your code on either a robot or in
+  simulation. If there's something broken in RobotPy, file an issue to get it
+  fixed
+* Format your code with black
+
+General:
+
+* We always try to stay as close to the original examples as possible
+* `Main.java` is never needed, it is equivalent to `wpilib.run(MyRobot)`
+* `robot.py` should have `#!/usr/bin/env python3` as the very first line
+* Don't ever check in files for your IDE (.vscode, .idea, etc)
+* Copy over the copyright statement from the original file
+
+Naming:
+
+* Filenames should always be all lowercase
+* Function names are camelCase
+* Class names start with a capital letter
+* Class method names are camelCase
+* Class member variables such as `m_name` should be `self.name` in Python
+* Protected/private methods/members can optionally be prefixed with `_`
+
+Misc conversion thoughts
+
+* Comparisons to null such as `foo == null` become `foo is None`
+* Single-line lambdas can be converted to python lambda statements. Anything
+  longer needs to be a separate function somewhere
+* Never modify `sys.path` directly!
+
+Longer thoughts
+---------------
+
+Never initialize anything other than constants at class/global scope. Here are
+a few examples:
+
+```python
+
+# OK: just a constant
+MY_CONSTANT = 42
+
+# BAD: at global scope
+motor = wpilib.Talon(1)
+
+class MyRobot:
+    # BAD: at class scope
+    motor = wpilib.Talon(1)
+
+    def __init__(self):
+        # OK: variable assigned to `self`
+        self.motor = wpilib.Talon(1)
+```
+
+---
+
+Import order doesn't really matter, but we prefer the following convention:
```

Comment:
Would it be worth setting up `isort` (or ruff) to do this? I'd suggest [`shed`] to automatically configure isort, but I'm not sure it'd work with our directory structure.

If we do it in [pre-commit], [pre-commit.ci] can do it for contributors.

[`shed`]: https://github.com/Zac-HD/shed
[pre-commit]: https://pre-commit.com
[pre-commit.ci]: https://pre-commit.ci

Reply:
If you want to do it, that's fine. I don't feel strongly enough about this guideline to enforce it -- but for those who read it, it's nice to keep in mind.